frame/election-provider-multi-phase/src/signed.rs

```diff
@@ -0,0 +1,707 @@
+		solution: RawSolution<CompactOf<T>>,
+		size: SolutionOrSnapshotSize,
+	) -> Option<usize> {
+		// Let's ensure that our input is valid.
+		let max_submissions = T::SignedMaxSubmissions::get();
+		debug_assert!(queue.len() as u32 <= max_submissions);
+
+		let threshold = T::SolutionImprovementThreshold::get();
+		// this insertion logic is a bit unusual in that a new solution which beats an existing
+		// solution by less than the threshold is sorted as "greater" than the existing solution.
+		// this means that the produced ordering depends on the order of insertion, and that
+		// attempts to produce a total ordering using this comparitor are highly unstable.
+		//
+		// this ordering prioritizes earlier solutions over slightly better later ones.
+		let insertion_position = queue.binary_search_by(|s| {
+			if is_score_better::<Perbill>(
+				solution.score,
+				s.solution.score,
+				threshold,
+			) {
+				Ordering::Greater
+			} else {
+				Ordering::Less
+			}
+		}).expect_err("comparitor function never returns Ordering::Equal; qed");
+
+		// if this solution is the worst one so far and the queue is full, don't insert
+		if insertion_position == queue.len() && queue.len() as u32 >= max_submissions {
+			return None;
+		}
+
+		// add to the designated spot. If the length is too much, remove one.
+		let reward = Self::reward_for();
+		let deposit = Self::deposit_for(&solution, size);
+		let submission =
+			SignedSubmission { who: who.clone(), deposit, reward, solution };
+		queue.insert(insertion_position, submission);
+
+		// Remove the weakest if queue is overflowing.
+		// This doesn't adjust insertion_position: in the event that it might have, we'd have short-
+		// circuited above.
+		if queue.len() as u32 > max_submissions {
+			if let Some(SignedSubmission { who, deposit, .. }) = queue.pop() {
+				let _remainder = T::Currency::unreserve(&who, deposit);
+				debug_assert!(_remainder.is_zero());
+			}
+		}
+		debug_assert!(queue.len() as u32 <= max_submissions);
+		Some(insertion_position)
+	}
+
+	/// The feasibility weight of the given raw solution.
+	pub fn feasibility_weight_of(
+		solution: &RawSolution<CompactOf<T>>,
+		size: SolutionOrSnapshotSize,
+	) -> Weight {
+		T::WeightInfo::feasibility_check(
+			size.voters,
+			size.targets,
+			solution.compact.voter_count() as u32,
+			solution.compact.unique_targets().len() as u32,
+		)
+	}
+
+	/// Collect sufficient deposit to store this solution this chain.
+	///
+	/// The deposit is composed of 3 main elements:
+	///
+	/// 1. base deposit, fixed for all submissions.
+	/// 2. a per-byte deposit, for renting the state usage.
+	/// 3. a per-weight deposit, for the potential weight usage in an upcoming on_initialize
+	pub fn deposit_for(
+		solution: &RawSolution<CompactOf<T>>,
+		size: SolutionOrSnapshotSize,
+	) -> BalanceOf<T> {
+		let encoded_len: BalanceOf<T> = solution.using_encoded(|e| e.len() as u32).into();
```

Comment:
would that `encoded_len` not work here?

Reply:
It would: 7ce5afb56 . Just needs a little type hinting.